### DISCUSSION TITLE
Does RTSP Pull work with Signed Policy?

### DISCUSSION THREAD

**Discussion by bchah** [OvenMediaLabs/OvenMediaEngine#1736] [Help] [answered]
Hello friends, I notice that when you pass a WebRTC playback request that uses SignedPolicy to a stream where an OriginMap is defined with an RTSP pull, it attaches the policy + signature arguments to the RTSP pull request, therefore receiving a 404 from the RTSP server.

In Server.xml I have configured the pull as such:

```
<Origins>
            <Origin>
                <Location>/live/mystream</Location>
                <Pass>
                    <Scheme>rtsp</Scheme>
                    <Urls><Url>myrtspserver.com:554/main</Url></Urls>
                </Pass>
            </Origin>
        </Origins>
```

And here we see that the policy and signature for WebRTC playback is being attached to the pull request:

```[2024-11-06 14:34:31.906] E [SPRtcSig-t3333:1143055] RtspcStream | rtspc_stream.cpp:310  | Rtsp server(rtsp://myrtspserver.com:554/main?p=eyJ1cmxfZXhwaXJlIjoxNzMwOTAzNjgxNjk2fQ&s=0yWyM454JlxHWf5ic9X1UhfQiZ0) rejected the describe request : 404(Stream Not Found)```

**Comment by bchah** [comment]
Just an update to confirm that if I remove the policy + signature arguments from the WebRTC playback request, RTSP Pull functions perfectly as expected :)

So is my other use case supported, using this with SignedPolicy?

**Reply by getroot** [reply, marked as answer]
Try using the ForwardQueryParams option. However, this is an unofficial feature that I made a long time ago, so I don't know if it still works. Check it out and let me know if it doesn't work.
```
<Origin>
    <Location>/live/mystream</Location>
    <Pass>
        <Scheme>rtsp</Scheme>
        <Urls><Url>myrtspserver.com:554/main</Url></Urls>
        <ForwardQueryParams>false</ForwardQueryParams>
    </Pass>
</Origin>
```

**Reply by bchah** [reply]
Thanks @getroot, that works for pulling the stream! However it does not stay persistent like we have the option for if we create it via the API. Is there anywhere that the _ForwardQueryParams_ argument can be used in the Create Stream API? How about the other properties available only in the API like _persistent, noInputFailoverTimeoutMs_ etc...? Hopefully we could use the Create Stream API or the XML configuration and be able to use a consistent set of parameters in both places, or at least being able to use the complete set of parameters in the API 🙏

**Reply by bchah** [reply]
Sorry, I see now that information is contained in the _Clustering_ section of the docs - with _UnusedStreamDeletionTimeout_ and _NoInputFailoverTimeout_ being the arguments I was looking for. Still it would be helpful if you can please help me figure out how to use _ForwardQueryParams_ with the API.

**Reply by getroot** [reply]
When using the Create Stream API, is there any reason to append a query string for SignedPolicy to the URL?

**Reply by bchah** [reply]
When creating the stream, no. However there could be arguments on the URL (policy, signature) when playing back with WebRTC. Perhaps streams created with the API handle this differently. I will do some more testing there.

**Reply by getroot** [reply]
It's a bit confusing. The following two are completely different mechanisms.

(1) Pulling while requesting RTSP URL to stream creation API
(2) Pulling RTSP from OriginMap if there is no stream inside when trying to play with WebRTC

In case (2), there may be a query string in the WebRTC URL to pass SignedPolicy, and this may be added to the RTSP URL up to the query string when generating the RTSP URL according to the OriginMap policy. This can be removed with the ForwardQueryParams option.

In case (1), you can just request a plain RTSP URL. ForwardQueryParams option is not needed in case (1).
